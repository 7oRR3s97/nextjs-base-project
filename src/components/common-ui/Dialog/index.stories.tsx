import React from 'react';
import { ComponentStory, ComponentMeta } from '@storybook/react';
import { within, fireEvent } from '@storybook/testing-library';

import DialogComponent from './index';

export default {
  title: 'CommonUI/Dialog',
  component: DialogComponent,
  argTypes: { onClose: { action: 'closed' }, handleBack: { action: 'back' } },
  decorators: [
    (Story) => (
      <div style={{ position: 'relative', width: '100%', height: '100%' }}>
        <Story />
      </div>
    ),
  ],
} as ComponentMeta<typeof DialogComponent>;

const Template: ComponentStory<typeof DialogComponent> = (args) => (
  <DialogComponent {...args} />
);

export const Dialog = Template.bind({});
Dialog.args = {
  open: true,
  children: 'dialog',
  dialogTitle: true,
  title: 'dialog title',
  subtitle: 'dialog subtitle',
  backButton: true,
  storybook: true,
};
Dialog.play = async ({ canvasElement }) => {
  const canvas = within(canvasElement);

  fireEvent.click(canvas.getByTestId('ArrowBackIcon'));
  fireEvent.click(canvas.getByTestId('CloseIcon'));
};
