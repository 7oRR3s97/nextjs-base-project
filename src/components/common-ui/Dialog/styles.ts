import { styled } from '@mui/material/styles';
import Box from '@mui/material/Box';
import Dialog from '@mui/material/Dialog';

export const DialogContainer = styled(Dialog)(({ theme }) => ({
  borderRadius: theme.spacing(3),
  '& .MuiPaper-root': {
    alignItems: 'center',
    justifyContent: 'center',
    borderRadius: theme.spacing(3),
    width: '90%',
    overflow: 'hidden',
    [theme.breakpoints.up('sm')]: {
      width: '80%',
      maxWidth: 500,
    },
    [theme.breakpoints.up('md')]: {
      width: '50%',
    },
  },
}));

export const Container = styled(Box)(({ theme }) => ({
  padding: theme.spacing(7),
  width: 'fill-available',
  overflowY: 'scroll',

  '&::-webkit-scrollbar': {
    background: 'transparent',
    height: '0',
    width: theme.spacing(2),
    '&-thumb': {
      background: theme.palette.primary.main,
      borderRadius: theme.spacing(3),
    },
  },
}));

export const DataContainer = styled(Box)(({ theme }) => ({
  padding: theme.spacing(0, 4, 4),
  display: 'grid',
  justifyItems: 'center',
}));
