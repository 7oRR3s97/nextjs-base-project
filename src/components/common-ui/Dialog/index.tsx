import React from 'react';

import type { DialogProps } from '@mui/material/Dialog';

import DialogTitle from './components/DialogTitle';
import { DialogContainer, Container, DataContainer } from './styles';

interface Props extends DialogProps {
  open: boolean;
  onClose(): void;
  dialogTitle?: boolean;
  title?: string;
  subtitle?: string;
  backButton?: boolean;
  handleBack?: () => void;
  storybook?: boolean;
}

const Dialog: React.FC<Props> = ({
  open,
  onClose,
  dialogTitle,
  title,
  subtitle,
  backButton,
  handleBack,
  children,
  storybook,
  ...props
}) => {
  const sx = storybook ? { position: 'unset' } : props.sx;
  const disablePortal = storybook ? true : props.disablePortal;
  return (
    <DialogContainer
      {...props}
      disablePortal={disablePortal}
      sx={sx}
      open={open}
      onClose={onClose}
    >
      <Container>
        {dialogTitle ? (
          <DialogTitle
            title={title}
            subtitle={subtitle}
            onClose={onClose}
            backButton={backButton}
            handleBack={handleBack}
          />
        ) : null}
        <DataContainer>{children}</DataContainer>
      </Container>
    </DialogContainer>
  );
};

export default Dialog;
