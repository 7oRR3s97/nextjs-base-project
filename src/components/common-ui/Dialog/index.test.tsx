import React from 'react';
import {
  render,
  screen,
  cleanup,
  fireEvent,
  waitFor,
  act,
} from 'helpers/test-utils';
import { axe, toHaveNoViolations } from 'jest-axe';
import { composeStories } from '@storybook/testing-react';
import Dialog from './index';

import * as DialogStories from './index.stories';

expect.extend(toHaveNoViolations);

describe('Dialog', () => {
  let open = true;
  let back = false;
  const handleOpen = (value: boolean) => {
    open = value;
  };

  const handleBack = (value: boolean) => {
    back = value;
  };

  beforeEach(() => {
    open = true;
  });

  afterEach(() => {
    cleanup();
    jest.clearAllTimers();
  });

  const stories = Object.values(composeStories(DialogStories)).map((Story) => {
    return {
      name: Story.storyName!,
      Story,
    };
  });

  it.each(stories)(
    'should have no accessibility error in $name story',
    async ({ Story }) => {
      const { container } = render(<Story />);
      const AxeResults = await axe(container);
      expect(AxeResults).toHaveNoViolations();
    }
  );

  it.each(stories)(
    'should run $name story play function and set open to false',
    async ({ Story }) => {
      act(() => {
        render(
          <Story
            data-testid="dialog-component"
            onClose={() => handleOpen(false)}
            handleBack={() => handleBack(true)}
          />
        );
      });

      await act(async () => {
        await Story.play({
          canvasElement: screen.getByTestId('dialog-component'),
        });
      });

      await waitFor(() => {
        expect(open).toBeFalsy();
        expect(back).toBeTruthy();
      });

      expect(screen.queryByTestId('dialog-component')).toMatchSnapshot();
    }
  );

  it('should not render the dialog without dialogTitle and then render it', async () => {
    open = false;

    const { rerender } = render(
      <Dialog
        data-testid="dialog-component"
        open={open}
        onClose={() => handleOpen(false)}
        dialogTitle={false}
      >
        dialog
      </Dialog>
    );

    const nonExistingDialog = screen.queryByTestId('dialog-component');
    expect(nonExistingDialog).toMatchSnapshot();
    expect(nonExistingDialog).not.toBeInTheDocument();

    handleOpen(true);
    expect(open).toBeTruthy();

    rerender(
      <Dialog
        data-testid="dialog-component"
        open={open}
        onClose={() => handleOpen(false)}
        dialogTitle={false}
      >
        dialog
      </Dialog>
    );

    const existingDialog = screen.queryByTestId('dialog-component');
    expect(existingDialog).toMatchSnapshot();
    expect(existingDialog).toBeInTheDocument();
  });

  it('should render the dialog with dialogTitle but without title, subtitle and back button', () => {
    render(
      <Dialog
        data-testid="dialog-component"
        open={open}
        onClose={() => handleOpen(false)}
        dialogTitle
        backButton={false}
      >
        dialog
      </Dialog>
    );

    const dialog = screen.queryByTestId('dialog-component');
    expect(dialog).toMatchSnapshot();
    expect(dialog).toBeInTheDocument();
  });

  it('should close the dialog', () => {
    render(
      <Dialog
        data-testid="dialog-component"
        open={open}
        onClose={() => handleOpen(false)}
        backButton
        dialogTitle
        title="dialog-title"
        subtitle="dialog-subtitle"
        handleBack={() => handleBack(true)}
      >
        dialog
      </Dialog>
    );

    fireEvent.click(screen.getByTestId('CloseIcon'));
    expect(open).toBeFalsy();
  });

  it('should go back on dialog', () => {
    render(
      <Dialog
        data-testid="dialog-component"
        open={open}
        onClose={() => handleOpen(false)}
        backButton
        dialogTitle
        title="dialog-title"
        subtitle="dialog-subtitle"
        handleBack={() => handleBack(true)}
      >
        dialog
      </Dialog>
    );

    fireEvent.click(screen.getByTestId('ArrowBackIcon'));
    expect(back).toBeTruthy();
  });
});
