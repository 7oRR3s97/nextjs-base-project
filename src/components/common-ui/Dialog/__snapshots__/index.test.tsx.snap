// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dialog should not render the dialog without dialogTitle and then render it 1`] = `null`;

exports[`Dialog should not render the dialog without dialogTitle and then render it 2`] = `
<div
  class="MuiModal-root MuiDialog-root css-1xdf76q-MuiModal-root-MuiDialog-root"
  data-testid="dialog-component"
  role="presentation"
>
  <div
    aria-hidden="true"
    class="MuiBackdrop-root css-yiavyu-MuiBackdrop-root-MuiDialog-backdrop"
    style="opacity: 1;"
  />
  <div
    data-test="sentinelStart"
    tabindex="0"
  />
  <div
    class="MuiDialog-container MuiDialog-scrollPaper css-hz1bth-MuiDialog-container"
    role="presentation"
    style="opacity: 1;"
    tabindex="-1"
  >
    <div
      aria-labelledby=":r2:"
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation24 MuiDialog-paper MuiDialog-paperScrollPaper MuiDialog-paperWidthSm css-lzee2o-MuiPaper-root-MuiDialog-paper"
      role="dialog"
    >
      <div
        class="MuiBox-root css-1xadayu"
      >
        <div
          class="MuiBox-root css-b4z479"
        >
          dialog
        </div>
      </div>
    </div>
  </div>
  <div
    data-test="sentinelEnd"
    tabindex="0"
  />
</div>
`;

exports[`Dialog should render the dialog with dialogTitle but without title, subtitle and back button 1`] = `
<div
  class="MuiModal-root MuiDialog-root css-1xdf76q-MuiModal-root-MuiDialog-root"
  data-testid="dialog-component"
  role="presentation"
>
  <div
    aria-hidden="true"
    class="MuiBackdrop-root css-yiavyu-MuiBackdrop-root-MuiDialog-backdrop"
    style="opacity: 1;"
  />
  <div
    data-test="sentinelStart"
    tabindex="0"
  />
  <div
    class="MuiDialog-container MuiDialog-scrollPaper css-hz1bth-MuiDialog-container"
    role="presentation"
    style="opacity: 1;"
    tabindex="-1"
  >
    <div
      aria-labelledby=":r3:"
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation24 MuiDialog-paper MuiDialog-paperScrollPaper MuiDialog-paperWidthSm css-lzee2o-MuiPaper-root-MuiDialog-paper"
      role="dialog"
    >
      <div
        class="MuiBox-root css-1xadayu"
      >
        <div
          class="MuiBox-root css-qoi7s5"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-cq0u38-MuiSvgIcon-root"
            data-testid="CloseIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
            />
          </svg>
        </div>
        <div
          class="MuiBox-root css-b4z479"
        >
          dialog
        </div>
      </div>
    </div>
  </div>
  <div
    data-test="sentinelEnd"
    tabindex="0"
  />
</div>
`;

exports[`Dialog should run Dialog story play function and set open to false 1`] = `
<div
  class="MuiModal-root MuiDialog-root css-e76r08-MuiModal-root-MuiDialog-root"
  data-testid="dialog-component"
  role="presentation"
>
  <div
    aria-hidden="true"
    class="MuiBackdrop-root css-yiavyu-MuiBackdrop-root-MuiDialog-backdrop"
    style="opacity: 1;"
  />
  <div
    data-test="sentinelStart"
    tabindex="0"
  />
  <div
    class="MuiDialog-container MuiDialog-scrollPaper css-hz1bth-MuiDialog-container"
    role="presentation"
    style="opacity: 1;"
    tabindex="-1"
  >
    <div
      aria-labelledby=":r1:"
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation24 MuiDialog-paper MuiDialog-paperScrollPaper MuiDialog-paperWidthSm css-lzee2o-MuiPaper-root-MuiDialog-paper"
      role="dialog"
    >
      <div
        class="MuiBox-root css-1xadayu"
      >
        <div
          class="MuiBox-root css-qoi7s5"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-cq0u38-MuiSvgIcon-root"
            data-testid="CloseIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
            />
          </svg>
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-3bmbr7-MuiSvgIcon-root"
            data-testid="ArrowBackIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M20 11H7.83l5.59-5.59L12 4l-8 8 8 8 1.41-1.41L7.83 13H20v-2z"
            />
          </svg>
          <h5
            class="MuiTypography-root MuiTypography-h5 css-1xkaumy-MuiTypography-root"
          >
            dialog title
          </h5>
          <h6
            class="MuiTypography-root MuiTypography-h6 css-1eadog5-MuiTypography-root"
          >
            dialog subtitle
          </h6>
        </div>
        <div
          class="MuiBox-root css-b4z479"
        >
          dialog
        </div>
      </div>
    </div>
  </div>
  <div
    data-test="sentinelEnd"
    tabindex="0"
  />
</div>
`;
